### DISCUSSION TITLE
Errors when reading subscripts and superscripts in a pdf?

### DISCUSSION THREAD

**Discussion by Oren-H** [py-pdf/pypdf#1253] [Q&A]
I am trying to extract text from a pdf document that contains several chemical formulas. These formulas contain subscripts and superscripts that PYPDF2 is not recognizing when I use the extractText() function. Instead, the characters are either being read as normal numbers or are preceded with '\n'. Is there anyway I can fix this issue, or are their other tools I can use to accurately read subscripts and superscripts? Any help would be very appreciated.

**Comment by MartinThoma** [comment]
Hi @Oren-H ,

This is a shortcoming of PyPDF2 at the moment. There is no support for this.

## Problem 1: Unclear expected results

### Order of superscripts and subscripts

For `H₂O` you would likely expect exactly that, but what would you expect for

![image](https://user-images.githubusercontent.com/1658117/185546177-78c1308a-4093-48ad-922e-08f36ea2bd45.png)

You might argue that you're fine with

> ²³⁵₉₂U

as well as with

> ₉₂²³⁵U

Writing both on the same level is not possible with plain text.

### Problem 2: Nested elements

I'm not sure how common this is in Chemistry, but in mathematics you can have nested elements. Most commonly are nested fractions:

![image](https://user-images.githubusercontent.com/1658117/185546723-266af6da-aae6-4c88-b705-5a8ce3da4c05.png)

Now we would need to decide on a syntax. Most likely it would be something similar to the math syntax of TeX ... but that's really unclear. And it would be extremely hard to parse that.

## Problem 2: How should PyPDF2 recognize superscripts / subscripts?

The PDF format does not necessarily represent this in a semantically meaningful way. Meaning we would need to be clever about how we extract it.

## How does PyPDF2 continue?

For the moment, I would say the math-part is completely out of scope. For simple subscripts / superscripts I would hope that at some point one of our contributors tackles them.

**Reply by Oren-H** [reply]
Thank you for the thorough response! Do you know of other python packages/tools that would allow me to solve this issue. It is integral to my project. Additionally, is anyone working on simply adding different parameters that can be added to the extractText() function that would provide some specificity on how to deal with these different scripts. For instance, I imagine that there could be a number of boolean parameters that would allow the coder to specify how they would like subscripts,superscripts, and other characters to be stored.

**Reply by MartinThoma** [reply]
> Do you know of other python packages/tools that would allow me to solve this issue.

No. I'm not a ware of any project (no matter of Python or not) that could recognize subscripts in PDF properly
